lib/autoupdate/agent/process.go

```diff
@@ -109,27 +117,67 @@ func (s SystemdService) Reload(ctx context.Context) error {
 	default:
 		s.Log.InfoContext(ctx, "Gracefully reloaded.", unitKey, s.ServiceName)
 	}
+	// monitor logs all relevant errors, so we filter for a few outcomes
+	err = s.monitor(ctx, initPID)
+	if errors.Is(err, context.DeadlineExceeded) {
+		return trace.Errorf("timed out while waiting for process to start")
+	}
+	if errors.Is(err, context.Canceled) {
+		return context.Canceled
+	}
+	if err != nil {
+		return trace.Errorf("failed to monitor process")
+	}
+	return nil
+}
+
+// monitor for a started, healthy process.
+// monitor logs all errors that should be displayed to the user.
+func (s SystemdService) monitor(ctx context.Context, initPID int) error {
+	ctx, cancel := context.WithTimeout(ctx, monitorTimeout)
+	defer cancel()
+	tickC := time.NewTicker(monitorInterval).C
+
 	if initPID != 0 {
 		s.Log.InfoContext(ctx, "Monitoring PID file to detect crashes.", unitKey, s.ServiceName)
-		err := s.monitor(ctx, initPID)
-		if errors.Is(err, context.DeadlineExceeded) {
-			return trace.Errorf("timed out while waiting for process to start")
+		err := s.monitorPID(ctx, initPID, tickC)
+		if err != nil {
+			s.Log.ErrorContext(ctx, "Error monitoring for crashing PID.", errorKey, err, unitKey, s.ServiceName)
+			return trace.Wrap(err)
 		}
+	}
+
+	s.Log.InfoContext(ctx, "Monitoring diagnostic socket to detect readiness.", unitKey, s.ServiceName)
+	client := debug.NewClient(s.SocketPath)
+	err := s.waitForSocketPresent(ctx, tickC, missingSocketTimeout)
+	if errors.Is(err, os.ErrNotExist) {
+		s.Log.WarnContext(ctx, "Socket appears to be disabled.", unitKey, s.ServiceName)
+		return nil
```

Comment:
This check is racy, we don't atomically rename the listening socket in place and we don't pass it across processes, the existing socket is unlinked and then a new one is created in its place, so this can randomly pass the test. I'm not entirely sure how to work around this sort of fundamental problem without changing how Teleport sets up its listening socket, however.

Reply:
> existing socket is unlinked and then a new one is created in its place

Does this happen after the new process is running? Or is it unlinked by the old process and relinked by the new process?

As long as the old socket does not return healthy after the new PID is stable, I could only return `nil` if the socket only returns dial errors.